File: app/src/main/java/com/yashovardhan99/healersdiary/fragments/PatientBillView.java
package com.yashovardhan99.healersdiary.fragments;


import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.google.firebase.Timestamp;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.ListenerRegistration;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;
import com.yashovardhan99.healersdiary.R;
import com.yashovardhan99.healersdiary.activities.MainActivity;
import com.yashovardhan99.healersdiary.objects.Patient;

import java.text.DateFormat;
import java.text.NumberFormat;
import java.util.Calendar;
import java.util.Objects;

import javax.annotation.Nullable;

/**
 * A simple {@link Fragment} subclass.
 */
public class PatientBillView extends DialogFragment implements com.google.firebase.firestore.EventListener<DocumentSnapshot>, View.OnClickListener {

    int billtype;
    Boolean detailed;
    String uid;
    Patient patient;
    TextView billText;
    DocumentReference documentReference;
    ListenerRegistration registration;
    com.google.firebase.firestore.EventListener<QuerySnapshot> healingListener = new com.google.firebase.firestore.EventListener<QuerySnapshot>() {
        @SuppressLint("SetTextI18n")
        @Override
        public void onEvent(@Nullable QuerySnapshot queryDocumentSnapshots, @Nullable FirebaseFirestoreException e) {
            int no = 0;
            assert queryDocumentSnapshots != null;
            for (DocumentSnapshot documentSnapshot : queryDocumentSnapshots) {
                no++;
                if (detailed) {
                    String date = DateFormat.getDateTimeInstance().format(documentSnapshot.getDate("Date"));
                    billText.setText(billText.getText() + "\n" + date);
                }
            }
            patient.setDue(patient.getRate() * no);
            billText.setText(billText.getText() + "\n" + getString(R.string.bill,
                    patient.getName(),
                    DateFormat.getDateInstance().format(Calendar.getInstance().getTime()),
                    NumberFormat.getCurrencyInstance().format(patient.getDue()),
                    FirebaseAuth.getInstance().getCurrentUser().getDisplayName()));
        }
    };


    public PatientBillView() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View RootView = inflater.inflate(R.layout.fragment_patient_bill_view, container, false);
        billText = RootView.findViewById(R.id.billText);
        Button share = RootView.findViewById(R.id.shareBill);
        Button close = RootView.findViewById(R.id.closeDialog);
        close.setOnClickListener(this);
        share.setOnClickListener(this);
        Bundle bundle = getArguments();
        uid = bundle.getString(MainActivity.PATIENT_UID);
        billtype = bundle.getInt(MainActivity.BILL_TYPE);
        FirebaseFirestore db = FirebaseFirestore.getInstance();
        documentReference = db.collection(MainActivity.USERS)
                .document(FirebaseAuth.getInstance().getUid())
                .collection("patients")
                .document(uid);
        registration = documentReference.addSnapshotListener(this);
        patient = new Patient();
        patient.setUid(uid);
        detailed = bundle.getBoolean(MainActivity.INCLUDE_LOGS);
        return RootView;
    }

    @Override
    public void onEvent(@Nullable DocumentSnapshot documentSnapshot, @Nullable FirebaseFirestoreException e) {
        assert documentSnapshot != null;
        patient.setName(documentSnapshot.getString("Name"));
        patient.setDue(Double.parseDouble(Objects.requireNonNull(documentSnapshot.get("Due")).toString()));
        patient.setRate(Double.parseDouble(Objects.requireNonNull(documentSnapshot.get("Rate")).toString()));
        Calendar thisMonth = Calendar.getInstance();
        thisMonth.set(Calendar.HOUR_OF_DAY, 0);
        thisMonth.clear(Calendar.MINUTE);
        thisMonth.clear(Calendar.SECOND);
        thisMonth.clear(Calendar.MILLISECOND);
        thisMonth.set(Calendar.DAY_OF_MONTH, 1);
        switch (billtype) {
            case R.id.allBillButton:
                billText.setText(getString(R.string.bill, patient.getName(),
                        DateFormat.getDateInstance().format(Calendar.getInstance().getTime()),
                        NumberFormat.getCurrencyInstance().format(patient.getDue()),
                        Objects.requireNonNull(FirebaseAuth.getInstance().getCurrentUser()).getDisplayName()));
                break;
            case R.id.thisMonthBillButton:
                registration.remove();
                registration = documentReference.collection("healings").orderBy("Date", Query.Direction.ASCENDING)
                        .whereGreaterThanOrEqualTo("Date", new Timestamp(thisMonth.getTime()))
                        .addSnapshotListener(healingListener);
                break;
            case R.id.lastMonthBillButton:
                Calendar lastMonth = Calendar.getInstance();
                lastMonth.add(Calendar.MONTH,-1);
                lastMonth.set(Calendar.DATE,1);
                lastMonth.set(Calendar.HOUR_OF_DAY,0);
                lastMonth.set(Calendar.MINUTE,0);
                lastMonth.set(Calendar.SECOND,0);
                lastMonth.set(Calendar.MILLISECOND,0);
                Log.d("THISMONTH",thisMonth.getTime().toString());
                Log.d("LASTMONTH",lastMonth.getTime().toString());
                registration.remove();
                registration = documentReference.collection("healings")
                        .whereGreaterThanOrEqualTo("Date", new Timestamp(lastMonth.getTime()))
                        .whereLessThan("Date", new Timestamp(thisMonth.getTime()))
                        .orderBy("Date", Query.Direction.ASCENDING)
                        .addSnapshotListener(healingListener);
                break;
        }
    }

    @Override
    public void onDetach() {
        registration.remove();
        healingListener = null;
        super.onDetach();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()){
            case R.id.shareBill:
                billText.setSelected(true);
                String text = billText.getText().toString();
                Intent share = new Intent(Intent.ACTION_SEND);
                share.putExtra(Intent.EXTRA_TEXT,text);
                share.setType("text/plain");
                startActivity(Intent.createChooser(share,"Send Bill"));
                break;
            case R.id.closeDialog:
                dismiss();
                break;
        }
    }
}

File: app/src/main/java/com/yashovardhan99/healersdiary/fragments/PatientBillView.kt
package com.yashovardhan99.healersdiary.fragments


import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.*
import com.yashovardhan99.healersdiary.R
import com.yashovardhan99.healersdiary.activities.MainActivity
import com.yashovardhan99.healersdiary.objects.Patient
import java.text.DateFormat
import java.text.NumberFormat
import java.util.*


/**
 * A simple [Fragment] subclass.
 */
class PatientBillView : DialogFragment(), com.google.firebase.firestore.EventListener<DocumentSnapshot>, View.OnClickListener {

    private var billtype: Int = 0
    private var detailed: Boolean = false
    private lateinit var uid: String
    private lateinit var patient: Patient
    private lateinit var billText: TextView
    private lateinit var documentReference: DocumentReference
    private lateinit var registration: ListenerRegistration
    @SuppressLint("SetTextI18n")
    private var healingListener: com.google.firebase.firestore.EventListener<QuerySnapshot>? = com.google.firebase.firestore.EventListener { queryDocumentSnapshots, _ ->
        var no = 0
        assert(queryDocumentSnapshots != null)
        for (documentSnapshot in queryDocumentSnapshots!!) {
            no++
            if (detailed) {
                val date = DateFormat.getDateTimeInstance().format(documentSnapshot.getDate("Date"))
                billText.text = billText.text.toString() + "\n" + date
            }
        }
        patient.due = patient.rate * no
        billText.text = billText.text.toString() + "\n" + getString(R.string.bill,
                patient.name,
                DateFormat.getDateInstance().format(Calendar.getInstance().time),
                NumberFormat.getCurrencyInstance().format(patient.due),
                FirebaseAuth.getInstance().currentUser!!.displayName)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val rootView = inflater.inflate(R.layout.fragment_patient_bill_view, container, false)
        billText = rootView.findViewById(R.id.billText)
        val share = rootView.findViewById<Button>(R.id.shareBill)
        val close = rootView.findViewById<Button>(R.id.closeDialog)
        close.setOnClickListener(this)
        share.setOnClickListener(this)
        val bundle = arguments
        uid = bundle!!.getString(MainActivity.PATIENT_UID)!!
        billtype = bundle.getInt(MainActivity.BILL_TYPE)
        val db = FirebaseFirestore.getInstance()
        documentReference = db.collection(MainActivity.USERS)
                .document(FirebaseAuth.getInstance().uid!!)
                .collection("patients")
                .document(uid)
        registration = documentReference.addSnapshotListener(this)
        patient = Patient()
        patient.uid = uid
        detailed = bundle.getBoolean(MainActivity.INCLUDE_LOGS)
        return rootView
    }

    override fun onEvent(documentSnapshot: DocumentSnapshot?, e: FirebaseFirestoreException?) {
        assert(documentSnapshot != null)
        patient.name = documentSnapshot!!.getString("Name")!!
        patient.due = java.lang.Double.parseDouble(documentSnapshot.get("Due")!!.toString())
        patient.rate = java.lang.Double.parseDouble(documentSnapshot.get("Rate")!!.toString())
        val thisMonth = Calendar.getInstance()
        thisMonth.set(Calendar.HOUR_OF_DAY, 0)
        thisMonth.clear(Calendar.MINUTE)
        thisMonth.clear(Calendar.SECOND)
        thisMonth.clear(Calendar.MILLISECOND)
        thisMonth.set(Calendar.DAY_OF_MONTH, 1)
        when (billtype) {
            R.id.allBillButton -> billText.text = getString(R.string.bill, patient.name,
                    DateFormat.getDateInstance().format(Calendar.getInstance().time),
                    NumberFormat.getCurrencyInstance().format(patient.due),
                    FirebaseAuth.getInstance().currentUser?.displayName)
            R.id.thisMonthBillButton -> {
                registration.remove()
                registration = documentReference.collection("healings").orderBy("Date", Query.Direction.ASCENDING)
                        .whereGreaterThanOrEqualTo("Date", Timestamp(thisMonth.time))
                        .addSnapshotListener(healingListener!!)
            }
            R.id.lastMonthBillButton -> {
                val lastMonth = Calendar.getInstance()
                lastMonth.add(Calendar.MONTH, -1)
                lastMonth.set(Calendar.DATE, 1)
                lastMonth.set(Calendar.HOUR_OF_DAY, 0)
                lastMonth.set(Calendar.MINUTE, 0)
                lastMonth.set(Calendar.SECOND, 0)
                lastMonth.set(Calendar.MILLISECOND, 0)
                Log.d("THISMONTH", thisMonth.time.toString())
                Log.d("LASTMONTH", lastMonth.time.toString())
                registration.remove()
                registration = documentReference.collection("healings")
                        .whereGreaterThanOrEqualTo("Date", Timestamp(lastMonth.time))
                        .whereLessThan("Date", Timestamp(thisMonth.time))
                        .orderBy("Date", Query.Direction.ASCENDING)
                        .addSnapshotListener(healingListener!!)
            }
        }
    }

    override fun onDetach() {
        registration.remove()
        healingListener = null
        super.onDetach()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.shareBill -> {
                billText.isSelected = true
                val text = billText.text.toString()
                val share = Intent(Intent.ACTION_SEND)
                share.putExtra(Intent.EXTRA_TEXT, text)
                share.type = "text/plain"
                startActivity(Intent.createChooser(share, "Send Bill"))
            }
            R.id.closeDialog -> dismiss()
        }
    }
}// Required empty public constructor
